Index: source/texk/web2c/luatexdir/luafontloader/fontforge/fontforge/parsettfatt.c
===================================================================
--- source/texk/web2c/luatexdir/luafontloader/fontforge/fontforge/parsettfatt.c	(revision 3963)
+++ source/texk/web2c/luatexdir/luafontloader/fontforge/fontforge/parsettfatt.c	(working copy)
@@ -4954,13 +4954,14 @@
     if ( glyphs==NULL )
 return;
     fseek(ttf,start+4,SEEK_SET);
-    for ( i=0; i<cnt; ++i ) if ( glyphs[i]<info->glyph_cnt && info->chars[ glyphs[i]]!=NULL ) {
-	val = (int16) getushort(ttf);
+    for ( i=0; i<cnt; ++i ) {
+      val = (int16) getushort(ttf);
+      offset = getushort(ttf);
+      if ( glyphs[i]<info->glyph_cnt && info->chars[ glyphs[i]]!=NULL ) {
 	if ( is_ic )
 	    info->chars[ glyphs[i] ]->italic_correction = val;
 	else
 	    info->chars[ glyphs[i] ]->top_accent_horiz = val;
-	offset = getushort(ttf);
 #ifdef FONTFORGE_CONFIG_DEVICETABLES
 	if ( offset!=0 ) {
 	    DeviceTable *dv = chunkalloc(sizeof(DeviceTable));
@@ -4971,6 +4972,7 @@
 		info->chars[ glyphs[i] ]->top_accent_adjusts = dv;
 	}
 #endif
+      }
     }
     free(glyphs);
 }
